File: apps/web/src/lib/api-client/teams.ts
import { apiRequestJson, apiRequestNoContent, pathSegment } from "./client";

type TeamMember = {
  userId: string;
  username: string;
  displayName: string;
  isManager: boolean;
  title: string;
  specialisms: string[];
  bio: string;
};

export async function listTeamMemberCandidates(
  teamId: string,
  query: string,
): Promise<{ users: TeamMember[] }> {
  return apiRequestJson<{ users: TeamMember[] }>(
    `/api/v1/teams/${pathSegment(teamId)}/member-candidates?query=${encodeURIComponent(query)}`,
    { method: "GET" },
  );
}

export type OrgTeam = {
  id: string;
  name: string;
  kind: "rfa" | "cm" | "jioc" | "qc";
  capabilityTeamId: string | null;
  members: TeamMember[];
};

export type CalendarEntry = {
  id: string;
  userId: string;
  date: string;
  status: "available" | "on_task" | "leave";
  note: string;
  createdByUserId: string | null;
};

export type TeamAvailability = {
  teamId: string;
  date: string;
  members: number;
  onLeave: number;
  onTaskCalendar: number;
  assignedLive: number;
  onTask: number;
  free: number;
};

export type UserProfile = {
  userId: string;
  title: string;
  specialisms: string[];
  bio: string;
  updatedAt: string;
};

export async function listTeams(): Promise<{ teams: OrgTeam[] }> {
  return apiRequestJson<{ teams: OrgTeam[] }>("/api/v1/teams", { method: "GET" });
}

export async function addTeamMember(
  teamId: string,
  userId: string,
  csrfToken: string,
): Promise<OrgTeam> {
  return apiRequestJson<OrgTeam>(`/api/v1/teams/${pathSegment(teamId)}/members`, {
    body: JSON.stringify({ userId }),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function removeTeamMember(
  teamId: string,
  userId: string,
  csrfToken: string,
): Promise<OrgTeam> {
  return apiRequestJson<OrgTeam>(
    `/api/v1/teams/${pathSegment(teamId)}/members/${pathSegment(userId)}`,
    { headers: { "X-CSRF-Token": csrfToken }, method: "DELETE" },
  );
}

export async function listTeamCalendar(
  teamId: string,
  from: string,
  to: string,
): Promise<{ entries: CalendarEntry[] }> {
  const query = `?from=${encodeURIComponent(from)}&to=${encodeURIComponent(to)}`;
  return apiRequestJson<{ entries: CalendarEntry[] }>(
    `/api/v1/teams/${pathSegment(teamId)}/calendar${query}`,
    { method: "GET" },
  );
}

export async function addCalendarEntry(
  teamId: string,
  payload: { userId: string; date: string; status: CalendarEntry["status"]; note?: string },
  csrfToken: string,
): Promise<CalendarEntry> {
  return apiRequestJson<CalendarEntry>(`/api/v1/teams/${pathSegment(teamId)}/calendar`, {
    body: JSON.stringify(payload),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function removeCalendarEntry(
  teamId: string,
  entryId: string,
  csrfToken: string,
): Promise<void> {
  await apiRequestNoContent(
    `/api/v1/teams/${pathSegment(teamId)}/calendar/${pathSegment(entryId)}`,
    { headers: { "X-CSRF-Token": csrfToken }, method: "DELETE" },
  );
}

export async function teamAvailability(teamId: string, date: string): Promise<TeamAvailability> {
  return apiRequestJson<TeamAvailability>(
    `/api/v1/teams/${pathSegment(teamId)}/availability?date=${encodeURIComponent(date)}`,
    { method: "GET" },
  );
}

export async function getMyProfile(): Promise<UserProfile> {
  return apiRequestJson<UserProfile>("/api/v1/users/me/profile", { method: "GET" });
}

export async function updateMyProfile(
  payload: { title: string; specialisms: string[]; bio: string },
  csrfToken: string,
): Promise<UserProfile> {
  return apiRequestJson<UserProfile>("/api/v1/users/me/profile", {
    body: JSON.stringify(payload),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "PUT",
  });
}

File: apps/api/src/coeus/api/presenters/teams.py
from coeus.domain.auth import UserAccount
from coeus.domain.teams import OrgTeam, TeamCalendarEntry, UserProfile, entry_end
from coeus.repositories.teams import TeamRepository
from coeus.schemas.teams import (
    AvailabilityResponse,
    CalendarEntryResponse,
    ProfileResponse,
    TeamMemberResponse,
    TeamResponse,
)
from coeus.services.team_availability import TeamAvailability


def team_response(
    team: OrgTeam, roster: tuple[UserAccount, ...], teams: TeamRepository
) -> TeamResponse:
    return TeamResponse(
        team_id=team.team_id,
        name=team.name,
        kind=team.kind.value,
        capability_team_id=team.capability_team_id,
        members=[_member_response(team, member, teams) for member in roster],
    )


def _member_response(
    team: OrgTeam, member: UserAccount, teams: TeamRepository
) -> TeamMemberResponse:
    profile = teams.get_profile(member.user_id)
    return TeamMemberResponse(
        user_id=member.user_id,
        username=member.username,
        display_name=member.display_name,
        is_manager=member.user_id in team.manager_user_ids,
        title=profile.title if profile else "",
        specialisms=list(profile.specialisms) if profile else [],
        bio=profile.bio if profile else "",
    )


def calendar_entry_response(entry: TeamCalendarEntry) -> CalendarEntryResponse:
    return CalendarEntryResponse(
        entry_id=entry.entry_id,
        user_id=entry.user_id,
        entry_date=entry.entry_date,
        end_date=entry_end(entry),
        status=entry.status.value,
        note=entry.note,
        created_by_user_id=entry.created_by_user_id,
    )


def availability_response(availability: TeamAvailability) -> AvailabilityResponse:
    return AvailabilityResponse(
        team_id=availability.team_id,
        entry_date=availability.entry_date,
        members=availability.members,
        on_leave=availability.on_leave,
        on_task_calendar=availability.on_task_calendar,
        other_commitments=availability.other_commitments,
        assigned_live=availability.assigned_live,
        on_task=availability.on_task,
        free=availability.free,
    )


def profile_response(profile: UserProfile) -> ProfileResponse:
    return ProfileResponse(
        user_id=profile.user_id,
        title=profile.title,
        specialisms=list(profile.specialisms),
        bio=profile.bio,
        updated_at=profile.updated_at,
    )
